### PR TITLE
Resize observer mixin: move to a reactive controller

### LINKED ISSUE #551: Rewrite the `withResizeObserver` mixin as a reactive controller
(title an description edited to reflect the latest proposition)

## Context

* We created the `withResizeObserver` mixin to hide the boilerplate of using `ResizeObserver` and provide nicers APIs for our components who need to be responsive
* Lit introduced the concept of [reactive controllers](https://lit.dev/docs/api/controllers/#ReactiveController)
  * This is clearly a better approach/solution than mixins
* The lit-labs project provides a [ready to use reactive controller](https://github.com/lit/lit/blob/main/packages/labs/observers/src/resize-controller.ts) to use the `ResizeObserver` API in a component

## About the [`ResizeController`](https://github.com/lit/lit/blob/main/packages/labs/observers/src/resize-controller.ts) from lit-labs

🙂 Pros:

* It doesn't do much things we don't need
* It seems very smart (just one `ResizeObserver` for many instances of the same component but multiple `observe` calls)

🤬 Cons:

* It doesn't provide the nice CSS API (with the automatic attributes on breakpoints) we have in our mixin
  * Most of those could be replaced by [CSS container queries](https://caniuse.com/css-container-queries) but the support is not good enough
* It just provides the `ResizeObserverEntry[]` but as we experienced it, it's not enough for us

## Proposition

* Create our own reactive controller reusing the code from our mixin
  * Try to see if we can reuse lit-labs's `ResizeController` or if we can get inspiration from it
* Replace all our mixin usage with the reactive controller
* Remove the mixin (is it a breaking change if it was not exposed / documented for public usage?)

WDYT?

### PR DESCRIPTION
**Pull request by florian-sanders-cc**
Fixes #551 

## Context

We decided to go with our own `resizeController` but with an API that does not rely on public properties:

- the controller should accept a `breakpoints` parameter,
- if `breakpoints != null` => it should add the relevant attributes,
- the controller should `requestUpdate` the `host` everytime a resize happens,
- if we need the `width` or `height` of the component, we can compute it within the `render` (or `willUpdate` if we need the size in several sub-renders).

## How to review?

- Review each commit,
- Review the [Resize Controller story](https://clever-components-preview.cellar-c2.services.clever-cloud.com/resize-observer/from-mixin-to-controller/index.html?path=/story/%F0%9F%95%B9%EF%B8%8F-controllers-resizecontroller--default-story),
- Review the [Resize Controller doc](https://clever-components-preview.cellar-c2.services.clever-cloud.com/resize-observer/from-mixin-to-controller/index.html?path=/docs/%F0%9F%95%B9%EF%B8%8F-controllers-resizecontroller--default-story),
- Review each impacted component using the [preview](https://clever-components-preview.cellar-c2.services.clever-cloud.com/resize-observer/from-mixin-to-controller/index.html), try to resize it and check that it does not break and it does not throw any error in the console:
  - cc-invoice-list
  - cc-invoice-table
  - cc-map
  - cc-orga-member-card
  - cc-overview
  - cc-pricing-product
  - cc-pricing-product-consumption
  - cc-tile-metrics
  - cc-tile-requests
  - cc-zone-input
- Don't forget to check the ADR! 

## TODO

- [x] Code the resize controller,
- [x] Provide a doc + a story for it,
- [x]  Replace `withResizeObserver` mixin with the `ResizeController` everywhere,
- [x] Write a short ADR about this.

<details>
<summary>Archived suggestions</summary>

In #551, we decided to rewrite the `resizeObserver` mixin and use a reactive controller instead.

The `resizeObserver` mixin has two features that we need:

- update the component internal `_size` prop with the current `width` of the component, everytime it is resized (we rely on a public `onResize` method to do so),
- set and remove attributes based on publicly declared `breakpoints` on the component everytime it is resized.

We want the `resizeController` to be able to do at least the same.

There are several ways to do so:

- building our own `resizeController` and using the component public methods and properties like we do with the mixin (first commit),
- relying on the lit-labs `resizeController` and using the component public methods and properties (second commit).
- relying on the lit-labs `resizeController` and using its API (`value` property of the observer + `willUpdate` lifecycle hook) (third commit).

All these options work, you can check the `cc-orga-member-card` / `cc-orga-member-list` (data loaded stories) since this component relies on both features (attribute & size prop).

</details>

### CODE REVIEW

File: src/components/cc-orga-member-card/cc-orga-member-card.js
Comment (florian-sanders-cc) [inline src/components/cc-orga-member-card/cc-orga-member-card.js:94]: using the `lit-labs` controller API, we no longer need to declare a public "protected" `breakpoints` property.

The controller will trigger the trigger the `willUpdate` lifecycle hook everytime a resize happens.

File: src/components/cc-orga-member-card/cc-orga-member-card.js
Comment (florian-sanders-cc) [inline src/components/cc-orga-member-card/cc-orga-member-card.js:102]: This is actually not mandatory.
Since we get the `width` of the component by using `element.getBoundingClientRect()`, we could just set up the controller with no callback and do:

```js
  this._size = this.getBoundingClientRect();
```

within `willUpdate` directly :thinking:.

**Comment by florian-sanders-cc** [comment]
We decided to go with our own `resizeController` but with an API that does not rely on public properties:

- the controller should accept a `breakpoints` parameter,
- if `breakpoints != null` => it should add the relevant attributes,
- the controller should `requestUpdate` the `host` everytime a resize happens,
- if we need the `width` or `height` of the component, we can compute it within the `render` (or `willUpdate` if we need the size in several sub-renders).

- [ ] TODO Write a short ADR about this

File: src/controllers/resize-controller.stories.js
Comment (pdesoyres-cc) [inline src/controllers/resize-controller.stories.js:36]: I have :frog: in my pond ;)

**Review by pdesoyres-cc — CHANGES_REQUESTED** [review]
Nice work @florian-sanders-cc

File: src/controllers/resize-controller.js
Comment (florian-sanders-cc) [reply src/controllers/resize-controller.js:42]: @pdesoyres-cc yes very good idea. This is actually what the Lit Labs implementation does but I don't know why I didn't go this way :thinking: 

Anyway, I'll let others review and suggest stuff but I'll probably go for this :+1: (and I'll also restore the comment as you suggest)

File: src/controllers/resize-controller.stories.js
Comment (florian-sanders-cc) [reply src/controllers/resize-controller.stories.js:36]: Frogs are cool too, I'll add it to the story :joy:

File: src/controllers/resize-controller.stories.js
Comment (HeleneAmouzou) [inline src/controllers/resize-controller.stories.js:17]: praise: great idea ! :smile_cat:

**Review by HeleneAmouzou — COMMENTED** [review]
GG @florian-sanders-cc ! 
There is just one small issue I can not solve: when I review the impacted components using the preview and try to resize it, sometimes the 350px size throw an error in the console ("Uncaught DOMException: Index or size is negative or greater than the allowed amount"), but as soon as I refresh the error is gone and when I switch again from other sizes to 350px it does not always re-throw it.

File: src/controllers/resize-controller.js
Comment (florian-sanders-cc) [reply src/controllers/resize-controller.js:35]: it's been forgotten, good catch :smile: thanks!

File: src/controllers/resize-controller.js
Comment (hsablonniere) [reply src/controllers/resize-controller.js:42]: Agreed with @pdesoyres-cc, I thought we decided that the ResizeController class would expose the `width` via a public getter like Lit does.

File: src/components/cc-invoice-table/cc-invoice-table.js
Comment (hsablonniere) [inline src/components/cc-invoice-table/cc-invoice-table.js:48]: issue: let's keep a reference to the controller so we can get its `width` property

```js
this._resizeController = new ResizeController(this);
```

File: src/components/cc-invoice-table/cc-invoice-table.js
Comment (hsablonniere) [inline src/components/cc-invoice-table/cc-invoice-table.js:57]: issue: let's use the controller reference for its width property:

```js
const bigMode = (this._resizeController.width > 700);
```

File: src/components/cc-map/cc-map.js
Comment (hsablonniere) [inline src/components/cc-map/cc-map.js:316]: question: I tried but I don't understand why it works compared to the `onResize () { this._map.invalidateSize(); }`

File: src/components/cc-orga-member-card/cc-orga-member-card.js
Comment (hsablonniere) [inline src/components/cc-orga-member-card/cc-orga-member-card.js:62]: issue: we don't need this

File: src/components/cc-orga-member-card/cc-orga-member-card.js
Comment (hsablonniere) [inline src/components/cc-orga-member-card/cc-orga-member-card.js:106]: issue: let's keep a reference to the controller so we can get its `width` property

```js
this._resizeController = new ResizeController(this);
```

File: src/components/cc-orga-member-card/cc-orga-member-card.js
Comment (hsablonniere) [inline src/components/cc-orga-member-card/cc-orga-member-card.js:230]: issue: we don't need this

File: src/components/cc-orga-member-card/cc-orga-member-card.js
Comment (hsablonniere) [inline src/components/cc-orga-member-card/cc-orga-member-card.js:234]: issue: we can use the width from the controller directly inside the render

```js
const width = this._resizeController.width;
```

File: src/components/cc-pricing-product-consumption/cc-pricing-product-consumption.js
Comment (hsablonniere) [inline src/components/cc-pricing-product-consumption/cc-pricing-product-consumption.js:94]: issue: let's keep a reference to the controller so we can get its `width` property

```js
this._resizeController = new ResizeController(this);
```

File: src/components/cc-pricing-product-consumption/cc-pricing-product-consumption.js
Comment (hsablonniere) [inline src/components/cc-pricing-product-consumption/cc-pricing-product-consumption.js:356]: issue: let's use the controller reference for its width property:

```js
const width = this._resizeController.width;
```

File: src/components/cc-pricing-product/cc-pricing-product.js
Comment (hsablonniere) [inline src/components/cc-pricing-product/cc-pricing-product.js:101]: issue: let's keep a reference to the controller so we can get its `width` property

```js
this._resizeController = new ResizeController(this);
```

File: src/components/cc-pricing-product/cc-pricing-product.js
Comment (hsablonniere) [inline src/components/cc-pricing-product/cc-pricing-product.js:261]: issue: let's use the controller reference for its width property:

```js
const width = this._resizeController.width;
```

File: src/components/cc-tile-metrics/cc-tile-metrics.js
Comment (hsablonniere) [inline src/components/cc-tile-metrics/cc-tile-metrics.js:253]: question: I tried but I don't understand why it works compared to the `onResize` + `this.*chart.resize()`

File: src/components/cc-tile-requests/cc-tile-requests.js
Comment (hsablonniere) [inline src/components/cc-tile-requests/cc-tile-requests.js:76]: issue: let's keep a reference to the controller so we can get its `width` property

```js
this._resizeController = new ResizeController(this);
```

File: src/components/cc-tile-requests/cc-tile-requests.js
Comment (hsablonniere) [inline src/components/cc-tile-requests/cc-tile-requests.js:224]: issue: let's use the controller reference for its width property:

```js
const width = this._resizeController.width;
```

File: src/controllers/resize-controller.stories.js
Comment (hsablonniere) [reply src/controllers/resize-controller.stories.js:17]: So nice! :heart:

File: docs/adr/adr-0024-from-resize-mixin-to-resize-controller.md
Comment (hsablonniere) [inline docs/adr/adr-0024-from-resize-mixin-to-resize-controller.md:262]: suggestion: can we add a link to container queries from MDN or something?

**Review by hsablonniere — CHANGES_REQUESTED** [review]
Nice work @florian-sanders-cc, especially the ADR :clap:

Like @pdesoyres-cc, I think the Controller should expose the `width`. I tried to annotate all files according for this change but you'll have to update the story, the doc and the ADR :pray:

It's clearly a nitpick but could you move the commit that removes the mixin at the end (or just before the ADR)?

---

FYI, even if we just read and not write, each time we call `getBoundingClientRect()`, we trigger a layout/reflow. It's always better not to call it too much so if the component asks the controller, it's better.

List of stuffs that trigger a layout/reflow:

https://gist.github.com/paulirish/5d52fb081b3570c81e3a

File: src/controllers/resize-controller.js
Comment (florian-sanders-cc) [reply src/controllers/resize-controller.js:71]: Good point! I'll change that, thanks :wink:

File: src/controllers/resize-controller.js
Comment (pdesoyres-cc) [inline src/controllers/resize-controller.js:51]: fix: Oh no, you don't want anybody but the controller itself to change the width. This is the whole point of making width private with only a getter on it. (This makes the width readonly from outside this class)

File: src/controllers/resize-controller.js
Comment (pdesoyres-cc) [inline src/controllers/resize-controller.js:57]: fix: use the private property directly
```suggestion
    this._width = this._host.getBoundingClientRect().width;
```

File: src/controllers/resize-controller.js
Comment (pdesoyres-cc) [inline src/controllers/resize-controller.js:77]: nitpick: you could use the internal property instead of the getter
```suggestion
this._callback(this._width);
```

File: src/controllers/resize-controller.js
Comment (pdesoyres-cc) [inline src/controllers/resize-controller.js:76]: nitpick: you can simplify that with :
```
this._callback?.(this._width);
```

**Review by pdesoyres-cc — CHANGES_REQUESTED** [review]
Hey, it's me again! We're nearly there.

File: src/controllers/resize-controller.js
Comment (florian-sanders-cc) [reply src/controllers/resize-controller.js:51]: ahhh yes of course :sweat_smile:  thanks

**Comment by florian-sanders-cc** [comment]
@pdesoyres-cc Thanks for your patience :joy: should be ok now? :eyes:

**Review by pdesoyres-cc — APPROVED** [review]
Well done Florian!

**Comment by florian-sanders-cc** [comment]
> There is just one small issue I can not solve: when I review the impacted components using the preview and try to resize it, sometimes the 350px size throw an error in the console ("Uncaught DOMException: Index or size is negative or greater than the allowed amount"), but as soon as I refresh the error is gone and when I switch again from other sizes to 350px it does not always re-throw it.

@HeleneAmouzou I couldn't reproduce this issue so I'll merge but when you are back we can test again together and if we still reproduce, we can open an issue and about this :+1:

**Pull request merged**
